numba/core/boxing.py

```diff
@@ -73,16 +73,26 @@ def box_float(typ, val, c):
 
 @unbox(types.Float)
 def unbox_float(typ, obj, c):
-    fobj = c.pyapi.number_float(obj)
-    dbval = c.pyapi.float_as_double(fobj)
-    c.pyapi.decref(fobj)
-    if typ == types.float32:
-        val = c.builder.fptrunc(dbval,
-                                c.context.get_argument_type(typ))
-    else:
-        assert typ == types.float64
-        val = dbval
-    return NativeValue(val, is_error=c.pyapi.c_api_error())
+    is_error_ptr = cgutils.alloca_once_value(c.builder, cgutils.false_bit)
+    float_val = cgutils.alloca_once(c.builder, c.context.get_argument_type(typ))
+    with ExitStack() as stack:
+        fobj = c.pyapi.number_float(obj)
+        with early_exit_if_null(c.builder, stack, fobj):
+            c.builder.store(cgutils.true_bit, is_error_ptr)
+
+        dbval = c.pyapi.float_as_double(fobj)
+        c.pyapi.decref(fobj)
+        with early_exit_if(c.builder, stack, c.pyapi.c_api_error()):
```

Comment:
This would be slightly faster as `if (dbval == -1 && PyErr_Occurred())`, but I have no idea how to write a short-circuiting `&&` in LLVM.